README
Gearman-Spawner version 1.10
============================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2011 by Say Media Inc

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.5 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Gearman::Spawner
1.10  Fri Aug  5 11:12:58 PDT 2011
    - Add unique support to sync client
    - Add known failure for background task in async client
1.09  Wed Jul 13 13:56:02 PDT 2011
    - Add unique support for background tasks
1.08  Mon May 23 16:15:13 PDT 2011
    - Pass 'unique' option through
1.07  Mon May 23 16:15:13 PDT 2011
    - Pass 'unique' option through
1.06 Tue May 17 23:37:50 PDT 2011
    - AnyEvent client: Make sure timeout doesn't fire after other callback has fired
1.05 Tue May 17 19:58:27 PDT 2011
    - Make client APIs similar and based on named-parameters
    - Document client interfaces
    - Move lists of common client tests to ClientTest lib
    - Test respawning behavior on die/exit/kill
    - Add timeout to AnyEvent client
    - Add tests for client timeout
1.04 Mon May 16 17:16:57 PDT 2011
    - Promote Server module to installation
1.03 Mon May 16 16:41:47 PDT 2011
    - Make client libs optional
1.02 Sun May 15 21:33:52 PDT 2011
    - Bind test server to localhost
1.01 Sun May 15 14:58:48 PDT 2011
    - Add verbosity for failed test port selection
1.00 Fri May 13 16:17:22 PDT 2011
    - Initial revision


t/lib/SlotWorker.pm
package SlotWorker;

use strict;
use warnings;

use base 'Gearman::Spawner::Worker';

sub new {
    my $self = shift;
    $self = fields::new($self) unless ref $self;
    $self->SUPER::new(@_);

    $self->register_method('slot');
    return $self;
}

sub slot {
    my SlotWorker $self = shift;
    sleep 1;
    return $self->{slot};
}

1;


t/lib/Gearman/Spawner/TestServer.pm
package Gearman::Spawner::TestServer;

use strict;
use warnings;

use Gearman::Spawner::Process;
use IO::Socket::INET;

use base 'Gearman::Spawner::Process';

my ($ADDRESS, $INSTANCE);
sub address {
    my $class = shift;

    return $ADDRESS if $ADDRESS;
    return $ADDRESS = $ENV{GEARMAN_SERVER} if $ENV{GEARMAN_SERVER};

    $INSTANCE = $class->create;
    return $ADDRESS = $INSTANCE->address;
}

sub create {
    my $class = shift;

    unless (eval "require Gearman::Server; 1") {
        die "need server to test against; either set GEARMAN_SERVER or install Gearman::Server\n";
    }

    return Gearman::Spawner::TestServer::Instance->new($class->fork_gearmand());
}

# start up a gearmand that exits when its parent process does. returns the
# address of the listening server and its pid
sub fork_gearmand {
    my $class = shift;

    # NB: this relies on Gearman::Spawner::Process allowing use of fork,
    # run_periodically, exit_with_parent, and loop as class methods instead of
    # object methods

    # get an unused port from the OS
    my $sock = IO::Socket::INET->new(
        Type      => SOCK_STREAM,
        Proto     => 'tcp',
        Reuse     => 1,
        LocalHost => 'localhost',
        Listen    => 1,
    );

    $sock or die "failed to request a listening socket: $!";

    my $port = $sock->sockport;
    my $host = $sock->sockhost;
    my $address = "$host:$port";
    $sock->close;

    my $parent_pid = $$;

    $Gearman::Spawner::Process::CHECK_PERIOD = 0.5;
    my $pid = $class->fork("[Gearman::Server $address] $0", 1);

    if ($pid) {
        # wait until server is contactable
        for (1 .. 50) {
            my $sock = IO::Socket::INET->new($address);
            return ($address, $pid) if $sock;
            select undef, undef, undef, 0.1;
        }
        die "couldn't contact server at $host:$port: $!";
    }

    require Gearman::Util; # Gearman::Server doesn't itself
    my $server = Gearman::Server->new;
    $server->create_listening_sock($port);

    $class->loop;
}

package Gearman::Spawner::TestServer::Instance;

use strict;
use warnings;

sub new {
    my $class = shift;
    my ($address, $pid) = @_;
    return bless {
        address => $address,
        pid => $pid,
        me => $$,
    }, $class;
}

sub address {
    my $self = shift;
    return $self->{address};
}

sub DESTROY {
    my $self = shift;
    kill 'INT', $self->{pid} if $$ == $self->{me};
}

1;


lib/Gearman/Spawner.pm
package Gearman::Spawner;

use strict;
use warnings;

use 5.10.0;

our $VERSION = '1.10';

use Carp qw( croak );
use Gearman::Spawner::Supervisor;

sub new {
    my $this = shift;
    my $class = ref $this || $this;

    my %params = @_;

    $params{servers} // croak "need servers";
    $params{workers} // croak "need workers";

    my $pid = Gearman::Spawner::Supervisor->start(%params);

    return bless {
        pid => $pid,
    }, $class;
}

sub pid {
    my $self = shift;
    return $self->{pid};
}

sub DESTROY {
    my $self = shift;
    kill 'INT', $self->{pid};
}

1;

__END__


=head1 NAME

Gearman::Spawner - Lifecycle manager for Gearman workers

=head1 SYNOPSIS

    # Launcher code
    $spawner = Gearman::Spawner->new(
        servers => ['localhost:7003'],
        preload => ['DBI'],
        workers => {
            MyWorker => {
                data     => {value => 3},
                max_jobs => 100,
            },
        },
    );


    # Worker class
    package MyWorker;
    use base 'Gearman::Spawner::Worker';

    sub new {
        my $class = shift;
        my $self = fields::new($class)->SUPER::new(@_);

        $self->register_method('mymethod');
        return $self;
    }

    sub mymethod {
        my MethodWorker $self = shift;
        my $arg = shift;
        return $arg + $self->{data}{value}; # value is 3, from above
    }


    # Client
    $client = Gearman::Spawner::Client::Sync->new(
        job_servers => \@servers,
    );
    my $retval = $client->run_method(MyWorker => mymethod => 4); # 7

=head1 DESCRIPTION

This class has a single method C<new> which creates a subprocess that manages
the lifecycles of configured Gearman workers. The workers must conform to a
simple OO interface, typically by inheriting from Gearman::Spawner::Worker.
The Gearman::Spawner object returned by C<new> may be queried by it's C<< ->pid
>> method to discover the identity of the process supervising the workers.
Additionally, when this object is destroyed, the spawner process and the child
workers are killed.

=head2 new

C<new> takes the following parameters:

=over 4

=item servers

(Required) A list of gearmand servers with which to register the workers.  Each
item in the list should be a string in the form "host:port".

=item preload

(Optional) A list of modules to be loaded before workers are forked. The memory
used by the loaded code can then be shared between the workers.

=item workers

(Required) A mapping of worker module names to the following configuration
structure:

=over 4

=item count

(Optional) The number of workers of this class to start. If not provided, 1 is
assumed.

=item data

(Optional) An opaque structure which will be made available to workers as an
object member.

=item max_jobs

(Optional) If specified, the number of jobs a worker will be allowed to
complete before exiting cleanly. The spawner will create a new worker to
replace it. Useful if your workers leak memory and should be recycled
occasionally.

=back

=back

=head1 FEATURES

Features:

=over 4

=item * All preloaded and worker modules will be tested for loadability before
workers are spawned so that failures are apparent to the caller (as an
exception).

=item * When the object is destroyed, the spawner process is killed, and so in
turn are the workers.

=item * If a worker dies unexpectedly, one will be created to replace it.

=item * Worker code requires just a simple constructor that registers one or
more methods; object method code then has access to both worker configuration
data and parameters of the job being executed.

=item * Serialization of client arguments and worker return values is handled
transparently.

=item * A synchronous client and a Danga::Socket-based asynchronous client are
included.

=item * Server-enforced timeouts for workers is supported.

=item * Methods are automatically namespaced according to worker class.

=back

=head1 AUTHOR

Adam Thomason, E<lt>athomason@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Say Media Inc, E<lt>cpan@sixapart.comE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/Gearman/Spawner/Client/Sync.pm
=head1 NAME

Gearman::Spawner::Client::Sync - synchronous client for Gearman::Spawner::Worker workers

=head1 SYNOPSIS

    $client = Gearman::Spawner::Client::Sync->new(
        job_servers => ['localhost:4730']
    );

    eval {
        my $result = $client->run_method(
            class  => 'MyWorker',
            method => 'sing',
            arg    => [qw( do re mi )],
        );
        say "success! result is $result";
    };
    if ($@) {
        say "failed because $@";
    }

=cut

package Gearman::Spawner::Client::Sync;

use strict;
use warnings;

use Gearman::Client;
use base 'Gearman::Client';

use Carp qw( croak );
use Gearman::Spawner::Util;
use Storable qw( nfreeze thaw );

=head1 METHODS

=over 4

=item Gearman::Spawner::Client::Sync->new(%options)

Creates a new client object. Options:

=over 4

=item job_servers

(Required) Arrayref of servers to connect to.

=back

=cut

sub new {
    my $ref = shift;
    my $class = ref $ref || $ref;

    my Gearman::Spawner::Client::Sync $self = fields::new($class)->SUPER::new(@_);

    return $self;
}

=item $client->run_method(%options)

Dispatches a foreground job to a worker.

Returns the deserialized result. If an error occurs, an exception is thrown.

Options:

=over 4

=item class

(Required) The name of the worker class.

=item method

(Required) The name of the method in I<class> to call.

=item data

(Optional) The job-specific data to pass to the worker. Any structure that can
be serialized with Storable is allowed. If omitted, undef is sent.

=item unique

(Optional) The opaque unique tag for coalescing jobs.

=back

=cut

sub run_method {
    my Gearman::Spawner::Client::Sync $self = shift;
    my %params = @_;

    my $class       = delete $params{class}         || croak "need class";
    my $method      = delete $params{method}        || croak "need method";
    my $data        = delete $params{data}          || undef;
    my $unique      = delete $params{unique}        || undef;

    croak "unknown parameters to run_method: @{[%params]}" if %params;

    my %options;
    $options{unique} = $unique if defined $unique;

    my $function = Gearman::Spawner::Util::method2function($class, $method);

    my $serialized = nfreeze([$data]);

    my $ref_to_frozen_retval = $self->do_task($function => $serialized, \%options);

    unless (defined $ref_to_frozen_retval) {
        die 'no return value from worker';
    }

    if (!ref $ref_to_frozen_retval || ref $ref_to_frozen_retval ne 'SCALAR') {
        die 'unexpected value type';
    }

    my $rets = eval { thaw($$ref_to_frozen_retval) };
    if ($@) {
        die "deserialization error: $@";
    }
    elsif (ref $rets ne 'ARRAY') {
        die "gearman function did not return an array";
    }

    return wantarray ? @$rets : $rets->[0];
}

=item run_method_background

Dispatches a background job to a worker.

Options:

=over 4

=item class

(Required) The name of the worker class.

=item method

(Required) The name of the method in I<class> to call.

=item data

(Optional) The job-specific data to pass to the worker. Any structure that can
be serialized with Storable is allowed. If omitted, undef is sent.

=item unique

(Optional) The opaque unique tag for coalescing jobs.

=back

=cut

sub run_method_background {
    my Gearman::Spawner::Client::Sync $self = shift;
    my %params = @_;

    my $class       = delete $params{class}         || croak "need class";
    my $method      = delete $params{method}        || croak "need method";
    my $data        = delete $params{data}          || undef;
    my $unique      = delete $params{unique}        || undef;

    croak "unknown parameters to run_method: @{[%params]}" if %params;

    my $function = Gearman::Spawner::Util::method2function($class, $method);

    my $serialized = nfreeze([$data]);

    my %options;
    $options{unique} = $unique if defined $unique;

    $self->dispatch_background($function => $serialized, \%options);

    return;
}

1;


lib/Gearman/Spawner/Client/Async.pm
=head1 NAME

Gearman::Spawner::Client::Async - asynchronous Danga::Socket client for Gearman::Spawner::Worker workers

=head1 SYNOPSIS

    $client = Gearman::Spawner::Client::Async->new(
        job_servers => ['localhost:4730']
    );

    $client->run_method(
        class  => 'MyWorker',
        method => 'sing',
        arg    => [qw( do re mi )],
        success_cb => sub {
            my $result = shift;
            say "success! result is $result";
        },
        error_cb => sub {
            my $reason = shift;
            say "failed because $reason";
        },
        timeout => 3,
    });

=cut

package Gearman::Spawner::Client::Async;

use strict;
use warnings;

use Gearman::Client::Async;
use base 'Gearman::Client::Async';

use Carp qw( croak );
use Gearman::Spawner::Util;
use Storable qw( nfreeze thaw );

=head1 METHODS

=over 4

=item Gearman::Spawner::Client::Async->new(%options)

Creates a new client object. Options:

=over 4

=item job_servers

(Required) Arrayref of servers to connect to.

=back

=cut

sub new {
    my $ref = shift;
    my $class = ref $ref || $ref;

    my Gearman::Spawner::Client::Async $self = fields::new($class)->SUPER::new(@_);
    return $self;
}

=item $client->run_method(%options)

Dispatches a foreground job to a worker. Options:

=over 4

=item class

(Required) The name of the worker class.

=item method

(Required) The name of the method in I<class> to call.

=item success_cb

(Required) The coderef to be called when the job completes successfully. The
first argument to it will be the deserialized result returned by the worker
method.

=item error_cb

(Required) The coderef to be called if the job does not complete. This may
occur for several reasons, including but not limited to: the worker code threw
an exception; the server did not respond before the timeout period; or the
number of job retries was exceeded.

The first argument passed to I<error_cb> is a string providing the best
available information about the error.

=item data

(Optional) The job-specific data to pass to the worker. Any structure that can
be serialized with Storable is allowed. If omitted, undef is sent.

=item timeout

(Optional) If the job has not completed or failed within this amount of time,
I<error_cb> will be called. Even if the job subsequently completes,
I<success_cb> will not be called.

=item unique

(Optional) The opaque unique tag for coalescing jobs.

=back

=cut

sub run_method {
    my Gearman::Spawner::Client::Async $self = shift;
    my %params = @_;

    my $class       = delete $params{class}         || croak "need class";
    my $method      = delete $params{method}        || croak "need method";
    my $success_cb  = delete $params{success_cb}    || croak "need success_cb";
    my $error_cb    = delete $params{error_cb}      || croak "need error_cb";
    my $data        = delete $params{data}          || undef;
    my $timeout     = delete $params{timeout}       || undef;
    my $unique      = delete $params{unique}        || undef;

    croak "unknown parameters to run_method: @{[%params]}" if %params;

    my $function = Gearman::Spawner::Util::method2function($class, $method);

    my $serialized = nfreeze([$data]);

    my %options;

    $options{timeout} = $timeout if defined $timeout;
    $options{unique}  = $unique  if defined $unique;

    $options{on_complete} = sub {
        my $ref_to_frozen_retval = shift;

        unless (defined $ref_to_frozen_retval) {
            return $error_cb->('no serialized return value from worker');
        }

        if (!ref $ref_to_frozen_retval || ref $ref_to_frozen_retval ne 'SCALAR') {
            return $error_cb->('unexpected value type');
        }

        my $rets = eval { thaw($$ref_to_frozen_retval) };
        if ($@) {
            return $error_cb->("deserialization error: $@");
        }
        elsif (ref $rets ne 'ARRAY') {
            return $error_cb->("gearman function did not return an array");
        }

        $success_cb->(@$rets);
    };

    $options{on_fail} = sub {
        my $reason = shift;
        $error_cb->($reason);
    };

    $self->add_task(Gearman::Task->new($function, \$serialized, \%options));
}

=item run_method_background

Dispatches a background job to a worker.

Options:

=over 4

=item class

(Required) The name of the worker class.

=item method

(Required) The name of the method in I<class> to call.

=item data

(Optional) The job-specific data to pass to the worker. Any structure that can
be serialized with Storable is allowed. If omitted, undef is sent.

=item unique

(Optional) The opaque unique tag for coalescing jobs.

=back

=cut

sub run_method_background {
    my $self = shift;
    my %params = @_;

    my $class       = delete $params{class}         || croak "need class";
    my $method      = delete $params{method}        || croak "need method";
    my $data        = delete $params{data}          || undef;
    my $unique      = delete $params{unique}        || undef;

    croak "unknown parameters to run_method_background: @{[%params]}" if %params;

    my %options;
    $options{unique}  = $unique  if defined $unique;

    my $function = Gearman::Spawner::Util::method2function($class, $method);

    my $serialized = nfreeze([$data]);

    # XXX dispatch_background does not exist in Gearman::Client::Async
    croak "dispatch_background is not supported by Async client";
    $self->dispatch_background($function => \$serialized, \%options);

    return;
}

1;


lib/Gearman/Spawner/Client/AnyEvent.pm
=head1 NAME

Gearman::Spawner::Client::AnyEvent - asynchronous AnyEvent client for Gearman::Spawner::Worker workers

=head1 SYNOPSIS

    $client = Gearman::Spawner::Client::AnyEvent->new(
        job_servers => ['localhost:4730']
    );

    $client->run_method(
        class  => 'MyWorker',
        method => 'sing',
        arg    => [qw( do re mi )],
        success_cb => sub {
            my $result = shift;
            say "success! result is $result";
        },
        error_cb => sub {
            my $reason = shift;
            say "failed because $reason";
        },
        timeout => 3,
    });

=cut

package Gearman::Spawner::Client::AnyEvent;

use strict;
use warnings;

use Any::Moose;

extends 'AnyEvent::Gearman::Client';

has cancel_timers => (
    is => 'rw',
    isa => 'HashRef',
);

no Any::Moose;

use AnyEvent;
use Gearman::Spawner::Util;

use Carp qw( croak );
use Storable qw( nfreeze thaw );

=head1 METHODS

=over 4

=item Gearman::Spawner::Client::AnyEvent->new(%options)

Creates a new client object. Options:

=over 4

=item job_servers

(Required) Arrayref of servers to connect to.

=back

=item $client->run_method(%options)

Dispatches a foreground job to a worker. Options:

=over 4

=item class

(Required) The name of the worker class.

=item method

(Required) The name of the method in I<class> to call.

=item success_cb

(Required) The coderef to be called when the job completes successfully. The
first argument to it will be the deserialized result returned by the worker
method.

=item error_cb

(Required) The coderef to be called if the job does not complete. This may
occur for several reasons, including but not limited to: the worker code threw
an exception; the server did not respond before the timeout period; or the
number of job retries was exceeded.

The first argument passed to I<error_cb> is a string providing the best
available information about the error.

=item data

(Optional) The job-specific data to pass to the worker. Any structure that can
be serialized with Storable is allowed. If omitted, undef is sent.

=item timeout

(Optional) If the job has not completed or failed within this amount of time,
I<error_cb> will be called. Even if the job subsequently completes,
I<success_cb> will not be called.

=item unique

(Optional) The opaque unique tag for coalescing jobs.

=back

=cut

sub run_method {
    my $self = shift;
    my %params = @_;

    my $class       = delete $params{class}         || croak "need class";
    my $method      = delete $params{method}        || croak "need method";
    my $success_cb  = delete $params{success_cb}    || croak "need success_cb";
    my $error_cb    = delete $params{error_cb}      || croak "need error_cb";
    my $data        = delete $params{data}          || undef;
    my $timeout     = delete $params{timeout}       || undef;
    my $unique      = delete $params{unique}        || undef;

    croak "unknown parameters to run_method: %params" if %params;

    my $function = Gearman::Spawner::Util::method2function($class, $method);

    my $serialized = nfreeze([$data]);

    my $timer;
    my $cancel_timeout;
    if (defined $timeout) {
        $cancel_timeout = sub {
            delete $self->{cancel_timers}{"$timer"};
            undef $timer;
        };
        $timer = AE::timer($timeout, 0, sub {
            $cancel_timeout->();
            $error_cb->("timeout");
        });
        $self->{cancel_timers}{"$timer"} = $timer;
    }

    my %options;

    $options{unique} = $unique if defined $unique;

    $options{on_complete} = sub {
        return if defined $timeout && !$timer; # timeout already fired
        $cancel_timeout->() if $cancel_timeout;

        my ($task, $frozen_retval) = @_;

        unless (defined $frozen_retval) {
            return $error_cb->('no serialized return value from worker');
        }

        my $rets = eval { thaw($frozen_retval) };
        if ($@) {
            return $error_cb->("deserialization error: $@");
        }
        elsif (ref $rets ne 'ARRAY') {
            return $error_cb->("gearman function did not return an array");
        }

        $success_cb->(@$rets);
    };

    $options{on_fail} = sub {
        return if defined $timeout && !$timer; # timeout already fired
        my ($task, $reason) = @_;
        $cancel_timeout->() if $cancel_timeout;
        $error_cb->($reason);
    };

    $self->add_task($function, $serialized, %options);
}

=item run_method_background

Dispatches a background job to a worker.

Options:

=over 4

=item class

(Required) The name of the worker class.

=item method

(Required) The name of the method in I<class> to call.

=item data

(Optional) The job-specific data to pass to the worker. Any structure that can
be serialized with Storable is allowed. If omitted, undef is sent.

=item unique

(Optional) The opaque unique tag for coalescing jobs.

=back

=cut

sub run_method_background {
    my $self = shift;
    my %params = @_;

    my $class       = delete $params{class}         || croak "need class";
    my $method      = delete $params{method}        || croak "need method";
    my $data        = delete $params{data}          || undef;
    my $unique      = delete $params{unique}        || undef;

    croak "unknown parameters to run_method_background: %params" if %params;

    my $function = Gearman::Spawner::Util::method2function($class, $method);

    my $serialized = nfreeze([$data]);

    my %options;

    $options{unique} = $unique if defined $unique;

    $self->add_task_bg($function => $serialized, %options);

    return;
}

=back

=cut

1;


lib/Gearman/Spawner/Server.pm
package Gearman::Spawner::Server;

use strict;
use warnings;

use Gearman::Spawner::Process;
use IO::Socket::INET;

use base 'Gearman::Spawner::Process';

my ($ADDRESS, $INSTANCE);
sub address {
    my $class = shift;

    return $ADDRESS if $ADDRESS;
    return $ADDRESS = $ENV{GEARMAN_SERVER} if $ENV{GEARMAN_SERVER};

    $INSTANCE = $class->create;
    return $ADDRESS = $INSTANCE->address;
}

sub create {
    my $class = shift;

    unless (eval "require Gearman::Server; 1") {
        die "need server to test against; either set GEARMAN_SERVER or install Gearman::Server\n";
    }

    return Gearman::Spawner::Server::Instance->new($class->fork_gearmand());
}

# start up a gearmand that exits when its parent process does. returns the
# address of the listening server and its pid
sub fork_gearmand {
    my $class = shift;

    # NB: this relies on Gearman::Spawner::Process allowing use of fork,
    # run_periodically, exit_with_parent, and loop as class methods instead of
    # object methods

    # get an unused port from the OS
    my $sock = IO::Socket::INET->new(
        Type      => SOCK_STREAM,
        Proto     => 'tcp',
        Reuse     => 1,
        LocalHost => 'localhost',
        Listen    => 1,
    );

    $sock or die "failed to request a listening socket: $!";

    my $port = $sock->sockport;
    my $host = $sock->sockhost;
    my $address = "$host:$port";
    $sock->close;

    my $parent_pid = $$;

    $Gearman::Spawner::Process::CHECK_PERIOD = 0.5;
    my $pid = $class->fork("[Gearman::Server $address] $0", 1);

    if ($pid) {
        # wait until server is contactable
        for (1 .. 50) {
            my $sock = IO::Socket::INET->new($address);
            return ($address, $pid) if $sock;
            select undef, undef, undef, 0.1;
        }
        die "couldn't contact server at $host:$port: $!";
    }

    require Gearman::Util; # Gearman::Server doesn't itself
    my $server = Gearman::Server->new;
    $server->create_listening_sock($port);

    $class->loop;
}

package Gearman::Spawner::Server::Instance;

use strict;
use warnings;

sub new {
    my $class = shift;
    my ($address, $pid) = @_;
    return bless {
        address => $address,
        pid => $pid,
        me => $$,
    }, $class;
}

sub address {
    my $self = shift;
    return $self->{address};
}

sub DESTROY {
    my $self = shift;
    kill 'INT', $self->{pid} if $$ == $self->{me};
}

1;
